system/lib/pthread/emscripten_futex_wait.c

```diff
@@ -143,7 +144,30 @@ int emscripten_futex_wait(volatile void *addr, uint32_t val, double max_wait_ms)
   if (max_wait_ms != INFINITY) {
     max_wait_ns = (int64_t)(max_wait_ms*1000*1000);
   }
+#ifdef __PIC__
+#define MAX_WAIT_NS (1000*1000*100)
+  // If dynamic linking is supported then we only wait for MAX_WAIT nanoseconds
+  // at a time so we can respond to dlopen calls that might happen on other
+  // threads.
```

Comment:
Rather than wait in small increments, could we wait normally and then do `_emscripten_process_dlopen_queue` when we wake up?

That is, I'm asking what is the benefit to finding out that there are dlopen events to handle early. It seems like we can handle them once per `futex_wait` but I'm probably missing something.

Reply:
Because we could be waiting forward.. for example, imagine a pool of workers, all asleep waiting for stuff to do.   

I suppose we could try to detect that case and somehow signal to the `dlopen` caller that either:
(a) All threads are not in sync
(b) All thread are either in sync, or currently sleeping and will get in sync on wakeup.

My intuition is that (b) would require some of additional locking, but perhaps you have some idea how it might work.   I'll see if I can come up with something.

Reply:
I managed to make this approach work!  Now workers can continue to sleep as long as they like.  Thanks for the suggestion @kripken !